README.md
40K-Battle-Generator
====================

These files will produce a web app for generating all the information needed to fight a tabletop battle.
It will cover all of the randomly-generated components which are needed.  It will aggregate all of the rules
for the game onto one, easily referenced page.  I plan on implementing varying degrees of randomness granularity, 
so that players using the web app can choose which components of the battle they'd like the system to generate and 
which components they'd rather produce themselves.  Eventually, I would like the battle generator to integrate into 
a map-based campaign tracker, so that players can click on a map territory to attack and have the system automatically
generate an appropriate battle.

I would like the battle generator to incorporate graphical interface components, with a minimal number of text input 
fields.  I would like to provide pop-up tooltips for special rules and screen elements so that all the content for a 
single battle can be contained on a single page without being overly cluttered.  I would like to use CSS3 and reactive
design so that the page is equally usable on a monitor or smart phone (It would be nice to be able to use the web app
while out and about.).  I plan on using JavaScript, with some add-on libraries to handle the application logic.  I'm
thinking about using XML to store data used by the generator, with a MySQL or Postgres database as an option for storing
long-term records, or handling the later campaign tracker addon.  I would like to add aesthetic components which capture
the feel of the setting, so that besides being functional, the generator will be stylistically pleasing.

This component plan may sound aggressive, but I plan on completing small functional pieces of it at a time incrementally,
so as to get a working prototype produced as quickly as possible, with refinements being produced insofar as the 
application seems to be useful.

40k/changes.txt
-Moved globals variable to 40k.js file
-

40k/scripts/fog.js
// Create an array to store our particles
var particles = [];

// The amount of particles to render
var particleCount = 30;

// The maximum velocity in each direction
var maxVelocity = 2;

// The target frames per second (how often do we want to update / redraw the scene)
var targetFPS = 33;

// Set the dimensions of the canvas as variables so they can be used.
var canvasWidth = 400;
var canvasHeight = 400;

// Create an image object (only need one instance)
var imageObj = new Image();

// Once the image has been downloaded then set the image on all of the particles
imageObj.onload = function() {
    particles.forEach(function(particle) {
            particle.setImage(imageObj);
    });
};

// Once the callback is arranged then set the source of the image
imageObj.src = "images/Smoke10.png";

// A function to create a particle object.
function Particle(context) {

    // Set the initial x and y positions
    this.x = 0;
    this.y = 0;

    // Set the initial velocity
    this.xVelocity = 0;
    this.yVelocity = 0;

    // Set the radius
    this.radius = 5;

    // Store the context which will be used to draw the particle
    this.context = context;

    // The function to draw the particle on the canvas.
    this.draw = function() {
        
        // If an image is set draw it
        if(this.image){
            this.context.drawImage(this.image, this.x-128, this.y-128);         
            // If the image is being rendered do not draw the circle so break out of the draw function                
            return;
        }
        // Draw the circle as before, with the addition of using the position and the radius from this object.
        this.context.beginPath();
        this.context.arc(this.x, this.y, this.radius, 0, 2 * Math.PI, false);
        this.context.fillStyle = "rgba(0, 255, 255, 1)";
        this.context.fill();
        this.context.closePath();
    };

    // Update the particle.
    this.update = function() {
        // Update the position of the particle with the addition of the velocity.
        this.x += this.xVelocity;
        this.y += this.yVelocity;

        // Check if has crossed the right edge
        if (this.x >= canvasWidth) {
            this.xVelocity = -this.xVelocity;
            this.x = canvasWidth;
        }
        // Check if has crossed the left edge
        else if (this.x <= 0) {
            this.xVelocity = -this.xVelocity;
            this.x = 0;
        }

        // Check if has crossed the bottom edge
        if (this.y >= canvasHeight) {
            this.yVelocity = -this.yVelocity;
            this.y = canvasHeight;
        }
        
        // Check if has crossed the top edge
        else if (this.y <= 0) {
            this.yVelocity = -this.yVelocity;
            this.y = 0;
        }
    };

    // A function to set the position of the particle.
    this.setPosition = function(x, y) {
        this.x = x;
        this.y = y;
    };

    // Function to set the velocity.
    this.setVelocity = function(x, y) {
        this.xVelocity = x;
        this.yVelocity = y;
    };
    
    this.setImage = function(image){
        this.image = image;
    }
}

// A function to generate a random number between 2 values
function generateRandom(min, max){
    return Math.random() * (max - min) + min;
}

// The canvas context if it is defined.
var context;

// Initialise the scene and set the context if possible
function init() {
    var canvas = document.getElementById('myCanvas');
    if (canvas.getContext) {

        // Set the context variable so it can be re-used
        context = canvas.getContext('2d');

        // Create the particles and set their initial positions and velocities
        for(var i=0; i < particleCount; ++i){
            var particle = new Particle(context);
            
            // Set the position to be inside the canvas bounds
            particle.setPosition(generateRandom(0, canvasWidth), generateRandom(0, canvasHeight));
            
            // Set the initial velocity to be either random and either negative or positive
            particle.setVelocity(generateRandom(-maxVelocity, maxVelocity), generateRandom(-maxVelocity, maxVelocity));
            particles.push(particle);            
        }
    }
    else {
        alert("Please use a modern browser");
    }
}

// The function to draw the scene
function draw() {
    // Clear the drawing surface and fill it with a black background
    context.fillStyle = "rgba(0, 0, 0, 0.5)";
    context.fillRect(0, 0, 400, 400);

    // Go through all of the particles and draw them.
    particles.forEach(function(particle) {
        particle.draw();
    });
}

// Update the scene
function update() {
    particles.forEach(function(particle) {
        particle.update();
    });
}

// Initialize the scene
init();

// If the context is set then we can draw the scene (if not then the browser does not support canvas)
if (context) {
    setInterval(function() {
        // Update the scene befoe drawing
        update();

        // Draw the scene
        draw();
    }, 1000 / targetFPS);
}

40k/scripts/battlefield.js
globals.addForts = false;	// Whether or not forts can be added to battlefield
globals.fortClicks = 0;
globals.forts = new Array();	// Information about battlefield fortifications
globals.fortLocations = new Array();
globals.terrain = new Array;



// Fog
var count = 6;
var spin = 1;
globals.placedSymbol = new Array();
fog = new Raster('fog');
fog.opacity = 0.2;
symbol = new Symbol(fog);

// Place fog on the map
for(var i = 0; i < count; i++)
{
	var center = Point.random()*view.size;
	globals.placedSymbol[i] = symbol.place(center);
	globals.placedSymbol[i].scale(5 * (i/count));
}

function onFrame(event)	// Animating the fog
{
	fog.rotate(spin);
}

globals.clear_fog = function()
{
	for(var j = 0; j < count; j++)
	{
		globals.placedSymbol[j].remove();	//Hallelujah!!!
	}
}
globals.addBattlefield = function(terrain)
{
	var field = new Raster(terrain);
	field.position = view.center -10;
}

/*
 *
 *	Function to draw deployment zones
 *
 */

globals.drawDeployment = function(faction_count, mission_type)
{
	var opacity = 0.30;
	var color1 = new RgbColor(1,0,0);
	var color2 = new RgbColor(0,0,1);
	var color3 = new RgbColor(0,1,0); 
	var color4 = new RgbColor(255,255,0);
	var size1 = new Size(60,360);
	var size2 = new Size(240,120);
		
	if(mission_type == 0)
	{		
		var zone1 = new Path.Circle(new Point(0,0), 60);
		zone1.fillColor = color1;
		zone1.strokeColor = color1;
		zone1.opacity = opacity;
		
		var zone2 = new Path.Circle(new Point(240, 240), 60);
		zone2.fillColor = color2;
		zone2.strokeColor = color2;
		zone2.opacity = opacity;
		
		if(faction_count == 4)
		{
			var zone3 = new Path.Circle(new Point(0,240), 60);
			zone3.fillColor = color3;
			zone3.strokeColor = color3;
			zone3.opacity = opacity;
			
			var zone4 = new Path.Circle(new Point(240,0), 60);
			zone4.fillColor = color4;
			zone4.strokeColor = color4;
			zone4.opacity = opacity;
			
			point1 = new Point(0,120);
			point2 = new Point(240,120);
			draw_center_line(point1,point2);
			
			point1 = new Point(120,0);
			point2 = new Point(120,240);
			draw_center_line(point1,point2);
			
		}
		else
		{
			point1 = new Point(240,0);
			point2 = new Point(0,240);
			draw_center_line(point1,point2);
		}
	}
	else if(mission_type == 1)
	{
		var roll = rollD3();
		switch(roll)
		{
			case 1:
				var zone1 = new Path.Rectangle(new Point(0,0),size1);
				zone1.fillColor = color1;
				zone1.strokeColor = color1;
				zone1.opacity = opacity;
				var zone2 = new Path.Rectangle(new Point(180, 0),size1);
				zone2.fillColor = color2;
				zone2.strokeColor = color2;
				zone2.opacity = opacity;
				point1 = new Point(120,0);
				point2 = new Point(120,360);
				draw_center_line(point1,point2);
				break;
			case 2:
				var zone1 = new Path.Rectangle(new Point(0,0),size2);
				zone1.fillColor = color1;
				zone1.strokeColor = color1;
				zone1.opacity = opacity;
				var zone2 = new Path.Rectangle(new Point(0,240),size2);
				zone2.fillColor = color2;
				zone2.strokeColor = color2;
				zone2.opacity = opacity;
				point1 = new Point(0,180);
				point2 = new Point(240,180);
				draw_center_line(point1,point2);
				break;
			case 3:
				var zone1 = new Path(new Point(0,0),new Point(170,0),new Point(0,255));
				zone1.closed = true;
				zone1.fillColor = color1;
				zone1.strokeColor = color1;
				zone1.opacity = opacity;
				
				var zone2 = new Path(new Point(70,360),new Point(240,105),new Point(240,360));
				zone2.closed = true;
				zone2.fillColor = color2;
				zone2.strokeColor = color2;
				zone2.opacity = opacity;
				
				point1 = new Point(240,0);
				point2 = new Point(0,360);
				draw_center_line(point1,point2);
				break;
		}
	}
	else
	{
		alert('special mission');
	}
	
	/*
	 *
	 *	Draw line(s) down the center of the battlefield
	 *
	 */
	
	function draw_center_line(point1,point2)
	{
		center_line = new Path.Line(point1,point2);
		center_line.strokeWidth = 4;
		center_line.strokeColor = 'white';
		center_line.dashArray = [10,4];
	}
}

globals.clearCanvas = function()	//Works!!!
{
	project.activeLayer.remove();
	var layer = new Layer();
}

 /**
 *
 *	Draw Terrain on map
 *
 */
 
globals.draw_terrain = function(id, x, y)
{
	var feature = globals.terrain.length;
	globals.terrain[feature] = new Raster(id);
	var rotation = Math.random()*360;
	globals.terrain[feature].position = new Point(x,y);
	globals.terrain[feature].rotate(rotation);
}

 /**
 *
 *	Clear Fortifications from the Map
 *
 */ 
 
globals.clearFort = function(number)
{
	globals.forts[number].remove();
	globals.density = new Array();	//	Tracking terrain density on the battlefield
	globals.density[0] = new Array();
	globals.density[1] = new Array();
	if(globals.game_size < 2)
	{
		globals.density[0] = [0,0];
		globals.density[1] = [0,0];
	}
	else
	{
		globals.density[0] = [0,0,0];
		globals.density[1] = [0,0,0];
	}
}


//	Handling mouse clicks on the canvas

function onMouseUp(event)
{
	if(globals.addForts == true)
	{
		//Track fort location
		var spot = new Point(event.point);
		
		/******************************
		For calculating terrain density
		******************************/
		var density_x = Math.floor(spot.x/120);
		var density_y = Math.floor(spot.y/120);
		globals.density[density_x][density_y] -= 1;
		/*****************************/
		
		spot = spot/60;
		spot.x = Math.floor(spot.x);
		spot.y = Math.floor(spot.y);
		globals.fortClicks++;
		globals.fortLocations[globals.fortClicks] = spot;
		document.getElementById("fort_count").innerHTML = String(globals.fortClicks);
		
		//Draw Fort on map
		spot.x = spot.x + 0.5;
		spot.y = spot.y + 0.5;
		globals.forts[globals.fortClicks] = new Raster('bastion');
		globals.forts[globals.fortClicks].position = new Point(spot*60);
		var rotation = Math.random()*360;
		globals.forts[globals.fortClicks].rotate(rotation);
	}
}

40k/scripts/40k.js
/**
 *
 *	TODO:	Establish Global Variables
 *
 */
 window.globals = {};	//Necessary for accessing Paper components via JavaScript
 globals.density = new Array();	//	Tracking terrain density on the battlefield
 globals.density[0] = new Array();
 globals.density[1] = new Array();
 globals.game_size;	//	How large the game is, to determine battlefield size
 var player_count;	//	The number of players in the game
 var faction_count;	//	How many sides are fighting in the battle (2 or 4);
 var objective_count; // How many objectives the scenario requires
 var mission_type;	// Small (0) Standard (1) or Special Mission (2)
 var mission_num;	// Which mission was selected
 var terrain_type;	// Which sort of battlefield is being fought on



/**
 *
 *	Testing
 *
 */ 

function test()
{
	console.log("(0,0): " + globals.density[0][0] + "(1,0): " + globals.density[1][0] + '\n'
	+ "(0,1): " + globals.density[0][1] + "(1,1): " + globals.density[1][1] + '\n'
	+ "(0,2): " + globals.density[0][2] + "(1,2): " + globals.density[1][2]);
}
 /**
 *
 *	Function to Initialize on page load
 *
 */ 
 
function init()
{
	add_tooltips();	// Add popup text across the page
	step1(); //	Start the battle generation process
}
 
/**
 *
 *	Step1: Ask for number of players
 *
 */
 
function step1()
{
	show("step1");
}
 
/**
 *
 *	Step2: Ask for number of factions
 *
 */
 
function step2(players)
{
	player_count = players;
	hide("step1");
	if(player_count == 4)
	{
		show("step2");
	}
	else
	{
		step3(2);
	}
	
}
 
 /**
 *
 *	Step3: Ask for game size
 *
 */
 
function step3(factions)
{
	faction_count = factions;
	hide("step2");
	show("step3");
	if(faction_count == 2)
	{
		var button = document.createElement('button');
		button.setAttribute('id','3button1');
		button.innerHTML='Less than 500 points';
		button.setAttribute('onclick','step4(0);');
		document.getElementById("step3").appendChild(button);
		
		button = document.createElement('button');
		button.setAttribute('id','3button2');
		button.innerHTML='500 points';
		button.setAttribute('onclick','step4(1);');
		document.getElementById("step3").appendChild(button);
		
		button = document.createElement('button');
		button.setAttribute('id','3button3');
		button.innerHTML='More than 500 points';
		button.setAttribute('onclick','step4(2);');
		document.getElementById("step3").appendChild(button);
	}
	else
	{
		var button = document.createElement('button');
		button.setAttribute('id','3button1');
		button.innerHTML='Less than 500 points';
		button.setAttribute('onclick','step4(0);');
		document.getElementById("step3").appendChild(button);
		
		button = document.createElement('button');
		button.setAttribute('id','3button2');
		button.innerHTML='500 points';
		button.setAttribute('onclick','step4(1);');
		document.getElementById("step3").appendChild(button);
	}
}
 
 /**
 *
 *	Step 4: Ask for Mission Type Preference
 *
 */

function step4(g_size)
{
	globals.game_size = g_size;
	hide("step3");
	show("step4");
	if(globals.game_size < 2)
	{
		shrink_battlefield();
		globals.density[0] = [0,0];
		globals.density[1] = [0,0];
	}
	else
	{
		globals.density[0] = [0,0,0];
		globals.density[1] = [0,0,0];
	}
	if(globals.game_size == 0)
	{
		show("suicide_mission");
		mission_num = 0;
		objective_count = 4;
		step5(0);
	}
	else if(globals.game_size == 1)
	{
		show("battleforce_recon");
		mission_num = 1;
		objective_count = 4;
		step5(0);
	}
}

 /**
 *
 *	TODO:	Step 5: Ask for Mission Preference
 *
 */
 
function step5(type)
{
	mission_type = type;
	hide("step4");
	switch(mission_type)
	{
		case 0:
			step6();
			break;
		case 1:
			show("step5a");
			break;
		case 2:
			show("step5b");
			break;
	}
}

 /**
 *
 *	TODO:	Step 6: Generate Random Mission and ask for Terrain Preference
 *
 */
 
function step6(number)
{
	hide("the_mission");
	hide("classified");
	if (number == 0)
	{
		number = rollD6();
	}
	switch(mission_type)
	{
		case 0:	// Small Mission
			break;
		case 1:	//Standard Mission
			hide("step5a");
			switch(number)
			{
				case 1:
					show('crusade');
					mission_num = 1;
					objective_count = rollD3() + 2;
					break;
				case 2:
					show('purge_the_alien');
					mission_num = 2;
					break;
				case 3:
					show('big_guns_never_tire');
					mission_num = 3;
					objective_count = rollD3() + 2;
					break;
				case 4:
					show('the_scouring');
					mission_num = 4;
					objective_count = 6;
					break;
				case 5:
					show('the_emperors_will');
					mission_num = 5;
					objective_count = 2;
					break;
				case 6:
					show('the_relic');
					mission_num = 6;
					break;
			}
			break;
		case 2:	// Special Mission
			hide("step5b");
			switch(number)
			{
				case 1:
					show('the_pact_of_blood');
					mission_num = 1;
					break;
				case 2:
					break;
				case 3:
					break;
				case 4:
					break;
				case 5:
					break;
				case 6:
					break;
			}
	}
	show("step6");	// Show terrain question
	
}

 /**
 *
 *	TODO:	Step 7: Generate Terrain and Deployment Zones
 *
 */
 
function step7(terrain)
{
	hide("step6")
	if(terrain=="random")
	{
		switch(rollD6())
		{
			case 1:
				terrain_type = "arctic";
				break;
			case 2:
				terrain_type = "plain";
				break;
			case 3:
				terrain_type = "urban";
				break;
			case 4:
				terrain_type = "overgrown";
				break;
			case 5:
				terrain_type = "rugged";
				break;
			case 6:
				terrain_type = "volcanic";
				break;
		}
	}
	else
	{
		terrain_type = terrain;
	}
	globals.clear_fog();
	globals.addBattlefield(terrain_type);
	globals.drawDeployment(faction_count, mission_type);
	show("step7");
	globals.addForts = true;	// Allow player to click on map
}

 /**
 *
 *	TODO:	Step8
 *
 */
 
 function step8()
 {
	globals.addForts = false;	//Keeps players from adding more forts
	hide("step7");
	show("step8");
	
	//Generate Terrain Density
	for(var i = 0; i < 2; i++)
	{
		if(globals.game_size < 2)
		{
			var height = 2;
		}
		else
		{
			var height = 3;
		}
		for (var j = 0; j < height; j++)
		{
			globals.density[i][j] += rollD3();
		}
	}
	
	//Draw Density Markers
	for(var i = 0; i < 2; i++)
	{
		if(globals.game_size < 2)
		{
			var height = 2;
		}
		else
		{
			var height = 3;
		}
		for (var j = 0; j < height; j++)
		{
			var number = globals.density[i][j];
			var x = (i*120)+60;
			var y = (j*120)+60;
			switch(number)
			{
				case 1:
					globals.draw_terrain('density_1',x,y);
					break;
				case 2:
					globals.draw_terrain('density_2',x,y);
					break;
				case 3:
					globals.draw_terrain('density_3',x,y);
					break;
				default:
					break;
			}
		}
	}
 }

 /**
 *
 *	Shrink Battlefield
 *
 */

function shrink_battlefield()
{
	document.getElementById("battlefield").setAttribute("height","240");
	document.getElementById("battlefield").setAttribute("style","margin-top: 25%;");
}

 /**
 *
 *	Clear Forts
 *
 */

function clear_forts()
{
	for(var i = 1; i < (globals.fortClicks+1); i++)
	{
		globals.clearFort(i);
	}
	globals.fortClicks = 0;
	globals.forts = new Array();
	document.getElementById("fort_count").innerHTML = "0";
}
 
 /**
 *
 *	TODO:	Step9
 *
 */
 
 /**
 *
 *	TODO:	Step10
 *
 */

/**
 *
 *	TODO:	Function to add tooltip to a piece of text
 *
 */
 
function add_tooltips()
{
	var tips = document.getElementsByTagName("span");
	for(var term, tip, i = 0; i < tips.length; i++)
	{
		term = tips.item(i);
		tip = term.getAttribute("tip");
		if(tip != "")
		{
			term.setAttribute("onmouseover","tooltip.show('"+tip+"');");
			term.setAttribute("onmouseout","tooltip.hide();");
			term.setAttribute("class","tip");
		}
	}
}
 
 /**
 *
 *	TODO:	Function to generate terrain densities
 *
 */
 
/**
 *
 *	TODO:	Function to assign terrain features
 *
 */ 

/**
 *
 *	TODO:	Function to generate deployment zones
 *
 */ 
 
/**
 *
 *	TODO:	Function to reset opacities on mouseover/click for each column
 *
 */ 
 
/**
 *
 *	Function to display hidden elements
 *
 */ 
 
function show(el_id)
{
	document.getElementById(el_id).style.display='inline';
}
 
 /**
 *
 *	Function to hide displayed elements
 *
 */ 
 
function hide(el_id)
{
	document.getElementById(el_id).style.display='none';
}
 
 
/**
 *
 *	Function to roll D6s
 *
 */
 
function rollD6(number)
{
	var total = 0;
	if(!number)
	{
		number = 1;
	}
	for (var i = 0; i < number; i++)
	{
		total = (total + Math.floor(Math.random()*6)+1);
	}
	return total;
}

/**
 *
 *	Function to roll D3s
 *
 */
 
function rollD3(number)
{
	var total = 0;
	if(!number)
	{
		number = 1;
	}
	for (var i = 0; i < number; i++)
	{
		total = (total + Math.floor(Math.random()*3)+1);
	}
	return total;
}

/**
*
* Tool tip function.  Add the stuff below to any text to get tooltip popups.
* The second argument is optional.
* onmouseover="tooltip.show('Testing  123 ', 200);"
* onmouseout="tooltip.hide();"
*/

var tooltip=function(){
	var id = 'tt';
	var top = 3;
	var left = -100;	// Originally 3
	var maxw = 200;	// Originally 300
	var speed = 30;
	var timer = 10;
	var endalpha = 95;
	var alpha = 0;
	var tt,t,c,b,h;
	var ie = document.all ? true : false;
	return{
		show:function(v,w){
			if(tt == null){
				tt = document.createElement('div');
				tt.setAttribute('id',id);
				t = document.createElement('div');
				t.setAttribute('id',id + 'top');
				c = document.createElement('div');
				c.setAttribute('id',id + 'cont');
				b = document.createElement('div');
				b.setAttribute('id',id + 'bot');
				tt.appendChild(t);
				tt.appendChild(c);
				tt.appendChild(b);
				document.body.appendChild(tt);
				tt.style.opacity = 0;
				tt.style.filter = 'alpha(opacity=0)';
				document.onmousemove = this.pos;
			}
			tt.style.display = 'block';
			c.innerHTML = v;
			tt.style.width = w ? w + 'px' : 'auto';
			if(!w && ie){
				t.style.display = 'none';
				b.style.display = 'none';
				tt.style.width = tt.offsetWidth;
				t.style.display = 'block';
				b.style.display = 'block';
			}
			if(tt.offsetWidth > maxw){tt.style.width = maxw + 'px'}
			h = parseInt(tt.offsetHeight) + top;
			clearInterval(tt.timer);
			tt.timer = setInterval(function(){tooltip.fade(1)},timer);
		},
		pos:function(e){
			var u = ie ? event.clientY + document.documentElement.scrollTop : e.pageY;
			var l = ie ? event.clientX + document.documentElement.scrollLeft : e.pageX;
			tt.style.top = (u - h) + 'px';
			tt.style.left = (l + left) + 'px';
		},
		fade:function(d){
			var a = alpha;
			if((a != endalpha && d == 1) || (a != 0 && d == -1)){
				var i = speed;
				if(endalpha - a < speed && d == 1){
					i = endalpha - a;
				}else if(alpha < speed && d == -1){
					i = a;
				}
				alpha = a + (i * d);
				tt.style.opacity = alpha * .01;
				tt.style.filter = 'alpha(opacity=' + alpha + ')';
			}else{
				clearInterval(tt.timer);
				if(d == -1){tt.style.display = 'none'}
			}
		},
		hide:function(){
			clearInterval(tt.timer);
			tt.timer = setInterval(function(){tooltip.fade(-1)},timer);
		}
	};
}();

/**
*
*	Sound Playing function
*
*/
function playSound(file)
{
	var snd = new Audio(file); // buffers automatically when created
	snd.play();
}